File: arch/server/test-useid.js
const React = require("react");
const ReactDOM = require("react-dom/server");

// Helper: component that renders a div with its useId value
function DivWithId() {
	const id = React.useId();
	return React.createElement("div", { id: id });
}

// Helper: component that calls useId twice
function DivWithTwoIds() {
	const id1 = React.useId();
	const id2 = React.useId();
	return React.createElement("div", { "data-id1": id1, "data-id2": id2 });
}

// Helper: component that calls useId three times
function DivWithThreeIds() {
	const id1 = React.useId();
	const id2 = React.useId();
	const id3 = React.useId();
	return React.createElement("div", {
		"data-id1": id1,
		"data-id2": id2,
		"data-id3": id3,
	});
}

// Helper: wrapper component (no useId)
function Wrapper({ children }) {
	return React.createElement("div", { className: "wrapper" }, children);
}

// Helper: component with useId that wraps children
function ParentWithId({ children }) {
	const id = React.useId();
	return React.createElement("div", { id: id }, children);
}

// Test 1: Single component with useId
function Test1() {
	return React.createElement(DivWithId);
}

// Test 2: Two sibling components with useId
function Test2() {
	return React.createElement(
		"div",
		null,
		React.createElement(DivWithId),
		React.createElement(DivWithId),
	);
}

// Test 3: Nested components with useId
function Test3() {
	return React.createElement(
		ParentWithId,
		null,
		React.createElement(DivWithId),
	);
}

// Test 4: Multiple useId calls in one component
function Test4() {
	return React.createElement(DivWithTwoIds);
}

// Test 5: Three useId calls in one component
function Test5() {
	return React.createElement(DivWithThreeIds);
}

// Test 6: Siblings with nested children
function Test6() {
	return React.createElement(
		"div",
		null,
		React.createElement(
			ParentWithId,
			null,
			React.createElement(DivWithId),
		),
		React.createElement(DivWithId),
	);
}

// Test 7: Deep nesting (3 levels)
function Test7() {
	return React.createElement(
		ParentWithId,
		null,
		React.createElement(
			ParentWithId,
			null,
			React.createElement(DivWithId),
		),
	);
}

// Test 8: Wrapper (no useId) doesn't affect IDs
function Test8() {
	return React.createElement(
		Wrapper,
		null,
		React.createElement(DivWithId),
	);
}

// Test 9: Three siblings
function Test9() {
	return React.createElement(
		"div",
		null,
		React.createElement(DivWithId),
		React.createElement(DivWithId),
		React.createElement(DivWithId),
	);
}

// Test 10: Sibling components each with nested children
function Test10() {
	return React.createElement(
		"div",
		null,
		React.createElement(
			ParentWithId,
			null,
			React.createElement(DivWithId),
			React.createElement(DivWithId),
		),
		React.createElement(
			ParentWithId,
			null,
			React.createElement(DivWithId),
		),
	);
}

// Test 11: Array/list of components (dynamic children via array)
function Test11() {
	const items = [0, 1, 2].map((i) =>
		React.createElement(DivWithId, { key: String(i) }),
	);
	return React.createElement("div", null, items);
}

// Test 12: Separate renderToString calls produce same IDs (reset per render)
function Test12a() {
	return React.createElement(DivWithId);
}
function Test12b() {
	return React.createElement(DivWithId);
}

const tests = [
	["Test 1: Single component with useId", Test1],
	["Test 2: Two sibling components", Test2],
	["Test 3: Nested components", Test3],
	["Test 4: Two useId calls in one component", Test4],
	["Test 5: Three useId calls in one component", Test5],
	["Test 6: Siblings with nested children", Test6],
	["Test 7: Deep nesting (3 levels)", Test7],
	["Test 8: Wrapper without useId", Test8],
	["Test 9: Three siblings", Test9],
	["Test 10: Complex siblings with nested", Test10],
	["Test 11: Array/list of components", Test11],
];

for (const [name, Component] of tests) {
	const html = ReactDOM.renderToString(React.createElement(Component));
	console.log(`${name}:`);
	console.log(`  ${html}`);
	console.log();
}

// Test 12: Separate renders
console.log("Test 12: Separate renders produce same IDs:");
const html12a = ReactDOM.renderToString(React.createElement(Test12a));
const html12b = ReactDOM.renderToString(React.createElement(Test12b));
console.log(`  render1: ${html12a}`);
console.log(`  render2: ${html12b}`);
console.log(`  same: ${html12a === html12b}`);

File: packages/reactDom/test/test_useId.ml
let assert_string left right = Alcotest.check Alcotest.string "should be equal" right left

(* Helper components *)
let div_with_id () =
  React.Upper_case_component
    ( "DivWithId",
      fun () ->
        let id = React.useId () in
        React.createElement "div" [ React.JSX.String ("id", "id", id) ] [] )

let div_with_two_ids () =
  React.Upper_case_component
    ( "DivWithTwoIds",
      fun () ->
        let id1 = React.useId () in
        let id2 = React.useId () in
        React.createElement "div"
          [ React.JSX.String ("data-id1", "data-id1", id1); React.JSX.String ("data-id2", "data-id2", id2) ]
          [] )

let div_with_three_ids () =
  React.Upper_case_component
    ( "DivWithThreeIds",
      fun () ->
        let id1 = React.useId () in
        let id2 = React.useId () in
        let id3 = React.useId () in
        React.createElement "div"
          [
            React.JSX.String ("data-id1", "data-id1", id1);
            React.JSX.String ("data-id2", "data-id2", id2);
            React.JSX.String ("data-id3", "data-id3", id3);
          ]
          [] )

let wrapper children =
  React.Upper_case_component
    ("Wrapper", fun () -> React.createElement "div" [ React.JSX.String ("class", "className", "wrapper") ] [ children ])

let parent_with_id children =
  React.Upper_case_component
    ( "ParentWithId",
      fun () ->
        let id = React.useId () in
        React.createElement "div" [ React.JSX.String ("id", "id", id) ] [ children ] )

let mk_provider ctx ~value ~children () = React.Context.provider ctx (React.Context.makeProps ~value ~children ())

(* All expected values verified against React 19.1.0 (bun arch/server/test-useid.js)
   React 19 ID format: \xc2\xab (U+00AB «) + prefix + R + treeId + \xc2\xbb (U+00BB ») *)

let single_component_with_use_id () =
  let html = ReactDOM.renderToString (div_with_id ()) in
  assert_string html "<div id=\"\xc2\xabR0\xc2\xbb\"></div>"

let two_sibling_components () =
  let el = React.createElement "div" [] [ div_with_id (); div_with_id () ] in
  let html = ReactDOM.renderToString el in
  assert_string html "<div><div id=\"\xc2\xabR1\xc2\xbb\"></div><div id=\"\xc2\xabR2\xc2\xbb\"></div></div>"

let nested_components () =
  let el = parent_with_id (div_with_id ()) in
  let html = ReactDOM.renderToString el in
  assert_string html "<div id=\"\xc2\xabR0\xc2\xbb\"><div id=\"\xc2\xabR1\xc2\xbb\"></div></div>"

let multiple_use_id_calls () =
  let html = ReactDOM.renderToString (div_with_two_ids ()) in
  assert_string html "<div data-id1=\"\xc2\xabR0\xc2\xbb\" data-id2=\"\xc2\xabR0H1\xc2\xbb\"></div>"

let three_use_id_calls () =
  let html = ReactDOM.renderToString (div_with_three_ids ()) in
  assert_string html
    "<div data-id1=\"\xc2\xabR0\xc2\xbb\" data-id2=\"\xc2\xabR0H1\xc2\xbb\" data-id3=\"\xc2\xabR0H2\xc2\xbb\"></div>"

let siblings_with_nested_children () =
  let el = React.createElement "div" [] [ parent_with_id (div_with_id ()); div_with_id () ] in
  let html = ReactDOM.renderToString el in
  assert_string html
    "<div><div id=\"\xc2\xabR1\xc2\xbb\"><div id=\"\xc2\xabR5\xc2\xbb\"></div></div><div \
     id=\"\xc2\xabR2\xc2\xbb\"></div></div>"

let deep_nesting () =
  let el = parent_with_id (parent_with_id (div_with_id ())) in
  let html = ReactDOM.renderToString el in
  assert_string html
    "<div id=\"\xc2\xabR0\xc2\xbb\"><div id=\"\xc2\xabR1\xc2\xbb\"><div id=\"\xc2\xabR3\xc2\xbb\"></div></div></div>"

let wrapper_without_use_id () =
  let el = wrapper (div_with_id ()) in
  let html = ReactDOM.renderToString el in
  assert_string html "<div class=\"wrapper\"><div id=\"\xc2\xabR0\xc2\xbb\"></div></div>"

let three_siblings () =
  let el = React.createElement "div" [] [ div_with_id (); div_with_id (); div_with_id () ] in
  let html = ReactDOM.renderToString el in
  assert_string html
    "<div><div id=\"\xc2\xabR1\xc2\xbb\"></div><div id=\"\xc2\xabR2\xc2\xbb\"></div><div \
     id=\"\xc2\xabR3\xc2\xbb\"></div></div>"

let complex_siblings_with_nested () =
  let el =
    React.createElement "div" []
      [
        parent_with_id (React.Fragment (React.List [ div_with_id (); div_with_id () ])); parent_with_id (div_with_id ());
      ]
  in
  let html = ReactDOM.renderToString el in
  assert_string html
    "<div><div id=\"\xc2\xabR1\xc2\xbb\"><div id=\"\xc2\xabRd\xc2\xbb\"></div><div \
     id=\"\xc2\xabRl\xc2\xbb\"></div></div><div id=\"\xc2\xabR2\xc2\xbb\"><div \
     id=\"\xc2\xabR6\xc2\xbb\"></div></div></div>"

let separate_renders_same_ids () =
  let html1 = ReactDOM.renderToString (div_with_id ()) in
  let html2 = ReactDOM.renderToString (div_with_id ()) in
  assert_string html1 html2

let static_markup_use_id () =
  let html = ReactDOM.renderToStaticMarkup (div_with_id ()) in
  assert_string html "<div id=\"\xc2\xabR0\xc2\xbb\"></div>"

let identifier_prefix () =
  let html = ReactDOM.renderToString ~identifier_prefix:"myapp" (div_with_id ()) in
  assert_string html "<div id=\"\xc2\xabmyappR0\xc2\xbb\"></div>"

(* ── Edge case tests (verified against React 19.1.0 output) ────────────────── *)

let use_id_inside_suspense () =
  let el =
    React.Suspense
      { key = None; children = div_with_id (); fallback = React.createElement "div" [] [ React.string "loading" ] }
  in
  let html = ReactDOM.renderToString el in
  assert_string html "<!--$--><div id=\"\xc2\xabR0\xc2\xbb\"></div><!--/$-->"

let use_id_suspense_and_sibling () =
  let el =
    React.createElement "div" []
      [
        React.Suspense
          { key = None; children = div_with_id (); fallback = React.createElement "div" [] [ React.string "loading" ] };
        div_with_id ();
      ]
  in
  let html = ReactDOM.renderToString el in
  assert_string html
    "<div><!--$--><div id=\"\xc2\xabR1\xc2\xbb\"></div><!--/$--><div id=\"\xc2\xabR2\xc2\xbb\"></div></div>"

let fragment_single_child () =
  let el = React.createElement "div" [] [ React.Fragment (div_with_id ()) ] in
  let html = ReactDOM.renderToString el in
  assert_string html "<div><div id=\"\xc2\xabR0\xc2\xbb\"></div></div>"

let fragment_multiple_children () =
  let el = React.createElement "div" [] [ React.Fragment (React.List [ div_with_id (); div_with_id () ]) ] in
  let html = ReactDOM.renderToString el in
  assert_string html "<div><div id=\"\xc2\xabR1\xc2\xbb\"></div><div id=\"\xc2\xabR2\xc2\xbb\"></div></div>"

let nested_fragments () =
  let el = React.createElement "div" [] [ React.Fragment (React.Fragment (div_with_id ())) ] in
  let html = ReactDOM.renderToString el in
  assert_string html "<div><div id=\"\xc2\xabR0\xc2\xbb\"></div></div>"

let null_between_siblings () =
  let el = React.createElement "div" [] [ div_with_id (); React.Empty; div_with_id () ] in
  let html = ReactDOM.renderToString el in
  assert_string html "<div><div id=\"\xc2\xabR1\xc2\xbb\"></div><div id=\"\xc2\xabR3\xc2\xbb\"></div></div>"

let many_siblings () =
  let children = List.init 10 (fun _ -> div_with_id ()) in
  let el = React.createElement "div" [] children in
  let html = ReactDOM.renderToString el in
  assert_string html
    "<div><div id=\"\xc2\xabR1\xc2\xbb\"></div><div id=\"\xc2\xabR2\xc2\xbb\"></div><div \
     id=\"\xc2\xabR3\xc2\xbb\"></div><div id=\"\xc2\xabR4\xc2\xbb\"></div><div id=\"\xc2\xabR5\xc2\xbb\"></div><div \
     id=\"\xc2\xabR6\xc2\xbb\"></div><div id=\"\xc2\xabR7\xc2\xbb\"></div><div id=\"\xc2\xabR8\xc2\xbb\"></div><div \
     id=\"\xc2\xabR9\xc2\xbb\"></div><div id=\"\xc2\xabRa\xc2\xbb\"></div></div>"

let provider_transparent () =
  let ctx = React.createContext "default" in
  let el = mk_provider ctx ~value:"provided" ~children:(div_with_id ()) () in
  let html = ReactDOM.renderToString el in
  assert_string html "<div id=\"\xc2\xabR0\xc2\xbb\"></div>"

let kitchen_sink () =
  let ctx = React.createContext "default" in
  let el =
    React.createElement "div" []
      [
        mk_provider ctx ~value:"a"
          ~children:
            (React.Fragment
               (React.List
                  [
                    div_with_id ();
                    React.Suspense
                      {
                        key = None;
                        children = div_with_id ();
                        fallback = React.createElement "span" [] [ React.string "..." ];
                      };
                  ]))
          ();
        div_with_id ();
      ]
  in
  let html = ReactDOM.renderToString el in
  assert_string html
    "<div><div id=\"\xc2\xabR5\xc2\xbb\"></div><!--$--><div id=\"\xc2\xabR9\xc2\xbb\"></div><!--/$--><div \
     id=\"\xc2\xabR2\xc2\xbb\"></div></div>"

let test title fn =
  (Printf.sprintf "ReactDOM.renderToString / useId / %s" title, [ Alcotest_lwt.test_case_sync "" `Quick fn ])

let tests =
  [
    test "single component with useId" single_component_with_use_id;
    test "two sibling components" two_sibling_components;
    test "nested components" nested_components;
    test "multiple useId calls in one component" multiple_use_id_calls;
    test "three useId calls in one component" three_use_id_calls;
    test "siblings with nested children" siblings_with_nested_children;
    test "deep nesting (3 levels)" deep_nesting;
    test "wrapper without useId is transparent" wrapper_without_use_id;
    test "three siblings" three_siblings;
    test "complex siblings with nested" complex_siblings_with_nested;
    test "separate renders produce same IDs" separate_renders_same_ids;
    test "renderToStaticMarkup also works" static_markup_use_id;
    test "identifier_prefix" identifier_prefix;
    test "useId inside Suspense (sync)" use_id_inside_suspense;
    test "Suspense with useId + sibling" use_id_suspense_and_sibling;
    test "Fragment single child is transparent" fragment_single_child;
    test "Fragment multiple children forks" fragment_multiple_children;
    test "Nested fragments transparent" nested_fragments;
    test "Null/Empty between siblings preserves slots" null_between_siblings;
    test "Many siblings (10, base-32 at Ra)" many_siblings;
    test "Provider is transparent" provider_transparent;
    test "Kitchen sink (Provider + Fragment + Suspense)" kitchen_sink;
  ]
